packages/labs/context/src/lib/controllers/context-provider.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {ContextRequestEvent} from '../context-request-event.js';
+import {ContextKey, ContextType} from '../context-key.js';
+import {ContextContainer} from './context-container.js';
+import {ReactiveController, ReactiveElement} from 'lit';
+
+declare global {
+  interface HTMLElementEventMap {
+    /**
+     * A 'context-provider' event can be emitted by any element which hosts
+     * a context provider to indicate it is available for use.
+     */
+    'context-provider': ContextProviderEvent<ContextKey<unknown, unknown>>;
+  }
+}
+
+export class ContextProviderEvent<
+  Context extends ContextKey<unknown, unknown>
+> extends Event {
+  /**
+   *
+   * @param context the context which this provider can provide
+   */
+  public constructor(public readonly context: Context) {
+    super('context-provider', {bubbles: true, composed: true});
+  }
+}
+
+/**
+ * A ReactiveController which can add context provider behavior to a
+ * custom-element.
+ *
+ * This controller simply listens to the `context-request` event when
+ * the host is connected to the DOM and registers the received callbacks
+ * against its observable Context implementation.
+ */
+export class ContextProvider<T extends ContextKey<unknown, unknown>>
+  extends ContextContainer<ContextType<T>>
+  implements ReactiveController
+{
+  constructor(
+    protected host: ReactiveElement,
+    private context: T,
+    initialValue?: ContextType<T>
+  ) {
+    super(initialValue);
+    this.host.addController(this);
+    this.attachListeners();
```

Comment:
Calling this in the constructor and in `hostConnected` doesn't make sense since it's double work. I guess the platform won't 2x add the listener, but the event will 2x dispatch, once potentially to outer space if the element is not connected at construction time.

It seems like the `request` event listener should be added here and the `provide` event should be dispatched in `hostConnected`.

Reply:
Hmm, I think you're right in that the `provide` event should only be dispatched in `hostConnected` but the event listener I think needs attachment in the constructor and in `hostConnected` since we remove the listener in disconnected?

Reply:
Have removed in 79da874 as it was unnecessary, tests seemed to confirm this too!